Simplex

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
 continuation of #25277

#### Brief description of what is fixed or changed

```console
Tools for optimizing a linear function for a given simplex.

For the linear objective function ``f`` with linear constraints
expressed using `Le`, `Ge` or `Eq` can be found with ``lpmin`` or
``lpmax``. The symbols are **unbounded** unless specifically
constrained.

As an alternative, the matrices describing the objective and the
constraints, and an optional list of bounds can be passed to
``linprog`` which will solve for the minimization of ``C*x``
under constraints ``A*x <= b`` and/or ``Aeq*x = beq``, and
individual bounds for variables given as ``(lo, hi)``. The values
returned are **nonnegative** unless bounds are provided that
indicate otherwise.

Errors that might be raised are UnboundedLPError when there is no
finite solution for the system or InfeasibleLPError when the
constraints represent impossible conditions (i.e. a non-existant
 simplex).

Here is a simple 1-D system: minimize `x` given that ``x >= 1``.

    >>> from sympy.solvers.simplex import lpmin, linprog
    >>> from sympy.abc import x

    The function and a list with the constraint is passed directly
    to `lpmin`:

    >>> lpmin(x, [x >= 1])
    (1, {x: 1})

    For `linprog` the matrix for the objective is `[1]` and the
    uivariate constraint can be passed as a bound with None acting
    as infinity:

    >>> linprog([1], bounds=(1, None))
    (1, [1])

    Or the matrices, corresponding to ``x >= 1`` expressed as
    ``-x <= -1`` as required by the routine, can be passed:

    >>> linprog([1], [-1], [-1])
    (1, [1])

    If there is no limit for the objective, an error is raised.
    In this case there is a valid region of interest (simplex)
    but no limit to how small ``x`` can be:

    >>> lpmin(x, [])
    Traceback (most recent call last):
    ...
    sympy.solvers.simplex.UnboundedLPError:
    Objective function can assume arbitrarily large values!

    An error is raised if there is no possible solution:

    >>> lpmin(x,[x<=1,x>=2])
    Traceback (most recent call last):
    ...
    sympy.solvers.simplex.InfeasibleLPError:
    Inconsistent/False constraint
```

#### Other comments
There are 3 public functions: lpmin, lpmax and linprog. 

`linprog` is patterned after the scipy interface though it has fewer options: 

```python
def linprog(c, A=None, b=None, A_eq=None, b_eq=None, bounds=None):
    """Return the minimization of ``c*x`` with the given
    constraints ``A*x <= b`` and ``A_eq*x = b_eq``. Unless bounds
    are given, variables will have nonnegative values in the solution.
```
The other two are for direct symbolic input (which may feel more natural). 

#### TODO
[x] I need to update the docstring to indicate the values are *not* constrained to be nonnegative when passing a matrix; the matrix is just a short-hand representation of the system without symbols.
[x] this doesn't work: `lpmin(x,[y >= 10, x >= y])`

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * simplex module added with public functions `lpmin`, `lpmax` and `linprog` for dealing with linear optimization
<!-- END RELEASE NOTES -->

orielmalihi committed:
linear programing function added

orielmalihi committed:
linear programing function added

orielmalihi committed:
docstring and tests for 'linear_programming' method added

orielmalihi committed:
docstring and tests for 'linear_programming' method added

orielmalihi committed:
deleted unused varaible and import

orielmalihi committed:
deleted unused varaible and import

orielmalihi committed:
added missing import

orielmalihi committed:
added missing import

orielmalihi committed:
added another missing import

orielmalihi committed:
added another missing import

orielmalihi committed:
fixed typos and added an example to the linear_programming method

orielmalihi committed:
fixed typos and added an example to the linear_programming method

orielmalihi committed:
removed unnecessary code and added tests to the linear_programming method

orielmalihi committed:
removed unnecessary code and added tests to the linear_programming method

orielmalihi committed:
removed unused import

orielmalihi committed:
removed unused import

orielmalihi committed:
rewrote the second example in the docstring

orielmalihi committed:
rewrote the second example in the docstring

orielmalihi committed:
rewrote the second example in the docstring.2

orielmalihi committed:
rewrote the second example in the docstring.2

orielmalihi committed:
removed unused import

orielmalihi committed:
removed unused import

orielmalihi committed:
refactoring the error messages

orielmalihi committed:
refactoring the error messages

orielmalihi committed:
created two unique custom error clssses

orielmalihi committed:
created two unique custom error clssses

orielmalihi committed:
added default value for D param in the linear_programming method

orielmalihi committed:
added default value for D param in the linear_programming method

orielmalihi committed:
removed __init__ and __str__ from custom exceptions

orielmalihi committed:
removed __init__ and __str__ from custom exceptions

orielmalihi committed:
Merge branch 'master' of git://github.com/sympy/sympy into ineq

orielmalihi committed:
after merging branch with master

orielmalihi committed:
after merging branch with master

orielmalihi committed:
updated AUTHORS file

orielmalihi committed:
updated AUTHORS file

orielmalihi committed:
removed 2 blank lines

orielmalihi committed:
removed 2 blank lines

orielmalihi committed:
updated AUTHORS file again

orielmalihi committed:
updated AUTHORS file again

orielmalihi committed:
Merge branch 'master' of git://github.com/sympy/sympy into ineq

orielmalihi committed:
shuffle method is now dereministic with option to choose its random seed

orielmalihi committed:
shuffle method is now dereministic with option to choose its random seed

orielmalihi committed:
Merge branch 'master' of git://github.com/sympy/sympy into ineq

orielmalihi committed:
Merge branch 'ineq' of https://github.com/orielmalihi/sympy into ineq

orielmalihi committed:
changed AUTHORS file

orielmalihi committed:
random is initialized out of while loop

orielmalihi committed:
Merge branch 'master' into ineq

orielmalihi committed:
updated docstring and removed test regarding symbols

orielmalihi committed:
removed unused imports

orielmalihi committed:
Merge branch 'ineq' of https://github.com/orielmalihi/sympy into ineq

orielmalihi committed:
removed blanked lines

orielmalihi committed:
changed Matrix to ImmutableMatrix

smichr committed:
Merge branch 'master' into ineq

TiloRC committed:
Merge remote-tracking branch 'orielmalihi/ineq' into simplex

TiloRC committed:
Linear programming: Take system of inequalities instead of matrices

The linear_programing function no longer takes four matrices as input.
Instead, now it takes a list of inequalities and an objective function.

TiloRC committed:
Apply suggestions from code review

Small improvements in formatting

Co-authored-by: Christopher Smith <smichr@gmail.com>

TiloRC committed:
Construct matrices by assigning variables to col in lexicographic order

TiloRC committed:
Adjust examples to match lexicographic order

TiloRC committed:
Round 1: Address comments from old pr (#21687)

TiloRC committed:
Add "See Also" to linear_programming

TiloRC committed:
Improve inequality to matrix func (_to_standard_form)

TiloRC committed:
Use core.random instead of passing seed as arg

TiloRC committed:
Merge branch 'master' into simplex

TiloRC committed:
Add to explainations of linear programming errors

TiloRC committed:
Fix spacing for doctest in _to_standard_form

TiloRC committed:
Add github email to mailmap

TiloRC committed:
Inequalities: Added find_feasible() function

Takes a list of linear inequalities and equalities as input and returns either
a list of variable values that satisfy the constraints or None if no feasible
solutions exist.

TiloRC committed:
Fixed doctest issues.

TiloRC committed:
Final round: Address remaining comments from old pr (sympy#21687)

This commit includes the following improvements:

- Streamlined _simplex for improved conciseness and efficiency.

- Add an option to skip the second phase of the simplex method, useful for the
new find_feasible function that only requires the first phase.

- Move column and row labeling management entirely within _simplex to avoid
duplicate code between find_feasible and linear_programming.

- Enhance docstrings for several functions.

- Add new test cases including test cases covering non-square matrices and
constraints with ">=" relations.

- Extend _to_standard_form to support AppliedBinaryRelation from the new
assumptions.

TiloRC committed:
Simplex method: use Bland's rule and improve code readability

TiloRC committed:
Remove seed from examples

TiloRC committed:
Linear programing: several improvements

- Simplex is now a non-internal method with a good docstring.

- Simplex takes A, B, C as required arguments and constructs M instead of just
taking M as its only required argument.

- Simplex: Raise exception if input contains numbers that are not floats or
Rational.

- Use SciPy to check optimum in test_linprog_maximize.

- Wrap docstring lines at 80 characters.
(Counting indents as characters which I did not do in a previous commit).

- Rename linear_programming to linprog_maximize.

TiloRC committed:
Add tests for Simplex and fix doctest error

TiloRC committed:
test_linprog_maximize: Fix optional dependency logic

TiloRC committed:
Make _simplex internal once again and add linprog_from_matricies

I also renamed linear_programing / linprog_maximize to
linprog_from_equations.

TiloRC committed:
Fix doctest

TiloRC committed:
Multiple changes

- fix spelling of matrices everywhere

- rename linprog_from_equations to linprog_maximize_from_equations

- add variables arguement to linprog_maximize_from_equations and
find_feasible

TiloRC committed:
Apply suggestions from code review

Minor spacing/formatting improvements

Co-authored-by: Christopher Smith <smichr@gmail.com>

TiloRC committed:
Move exception raising and remove float('inf')

- Matrix shape exception moved from _simplex to linprog_from_matrices

- In _choose_pivot_row initialize min_ratio with first row ratio
instead of using float('inf').

TiloRC committed:
Merge branch 'simplex' of https://github.com/TiloRC/sympy into simplex

TiloRC committed:
Fix 'see also', improve _pivot docstring, use tupples in R and S

TiloRC committed:
Use False and True instead of 'x' and 'y' for x and y variables

TiloRC committed:
Apply suggestions from code review

- simplify _pivot docstring explanation
- linprog_from_matrices and linprog_maximize_from_equations: improve Returns docstring section

Co-authored-by: Christopher Smith <smichr@gmail.com>

smichr committed:
simplify _pivot docstring

TiloRC committed:
Update inequalities.py

Fix example to pass doctests

TiloRC committed:
Changed wording: Ferguson's paper --> Ferguson's introduction

TiloRC committed:
wording: simplex -> two phase simplex

changed wording in docstring for _simplex

TiloRC committed:
Delete find_feasible

Code was moved to PR #25363. It's not possible to handle strict
inequalities using the version of the simplex method used by
_simplex.

smichr committed:
modify r3 so canonical form is different

smichr committed:
make relationals canonical

smichr committed:
Update sympy/solvers/inequalities.py

smichr committed:
Update sympy/solvers/inequalities.py

smichr committed:
Update sympy/solvers/inequalities.py

TiloRC committed:
Stict --> Strict

Co-authored-by: Christopher Smith <smichr@gmail.com>

TiloRC committed:
linprog testing: test randomly generated problems

TiloRC committed:
linprog_maximize_from_equations: don't ignore const term in obj

TiloRC committed:
Renamed functions and changed their signatures

_simplex(A, B, C) --> _simplex(A, B, C, D=ImmutableMatrix([0]))

linprog_from_matrices(A, B, C) -->
   linear_optimization_from_matrices(f, A, b)

linprog_maximize_from_equations(
   constraints, objective, variables) -->
   linear_optimization(f, cons, variables)

Also changed argmax_dual --> argmin_dual

TiloRC committed:
spelling: sparcity --> sparsity

Co-authored-by: Christopher Smith <smichr@gmail.com>

TiloRC committed:
Merge branch 'simplex' of https://github.com/TiloRC/sympy into simplex

TiloRC committed:
fix duplicate make_random_problem function

TiloRC committed:
is True --> == True

TiloRC committed:
make_random_problem: fix return arg order

smichr committed:
new proposals

smichr committed:
move to simplex module

smichr commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,617 @@
+
+from sympy.core import sympify
+from sympy.core.exprtools import factor_terms
+from sympy.core.relational import Le, Ge, Eq, Le
```
```suggestion
from sympy.core.relational import Le, Ge, Eq

```

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -41,13 +40,13 @@ def test_solve_poly_inequality():
+    assert reduce__inequalities(
         [[Lt(x**2, 0)]], x, relational=False) == S.EmptySet
-    assert reduce_rational_inequalities(
+    assert reduce__inequalities(
```
```suggestion
    assert reduce_rational_inequalities(
```

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -41,13 +40,13 @@ def test_solve_poly_inequality():
+    assert reduce__inequalities(
         [[Le(x**2, 0)]], x, relational=False) == FiniteSet(0)
-    assert reduce_rational_inequalities(
+    assert reduce__inequalities(
```
```suggestion
    assert reduce_rational_inequalities(
```

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -41,13 +40,13 @@ def test_solve_poly_inequality():
+    assert reduce__inequalities(
         [[Eq(x**2, 0)]], x, relational=False) == FiniteSet(0)
-    assert reduce_rational_inequalities(
+    assert reduce__inequalities(
```
```suggestion
    assert reduce_rational_inequalities(
```

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -41,13 +40,13 @@ def test_solve_poly_inequality():
 
 def test_reduce_poly_inequalities_real_interval():
-    assert reduce_rational_inequalities(
+    assert reduce__inequalities(
```
```suggestion
    assert reduce_rational_inequalities(
```

smichr commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,618 @@
+    """
+    r = {}  # replacements to handle change of variables
+    np = []  # nonpositive expressions
+    aux = []  # will contain all symbols when done
```
```suggestion
    aux = []  # auxilliary symbols added

```

smichr commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,618 @@
+    np = []  # nonpositive expressions
+    aux = []  # will contain all symbols when done
+    ui = numbered_symbols('z', start=1, cls=Dummy) # auxilliary symbols
+    univariate = {}  # bound to be inferred from univariate constraints
```
```suggestion
    univariate = {}  # {x: interval} for univariate constraints

```

smichr commented:
I'm having doubts that this even belongs in SymPy: it feels like feature creep. SymPy is primarily for things that need symbolic representation. The simplex method added here is strictly for concrete matrices with Float/Rational args. The work that Tiloc is doing with satisfiablility seems more of interest. Solving systems using the Fourier-Motzkin approach that was part of the original #22389 seems more SymPy apropos:
```python
    >>> symbols = {x, y, z}
    >>> eq1 = 2*x - 3*y + z + 1 > 0
    >>> eq2 = x - y + 2*z - 2 > 0
    >>> eq3 = x - y + 3*z + 4 > 0
    >>> eq4 = x - z > 0
    >>> d = _fourier_motzkin_extension([eq1, eq2, eq3, eq4], symbols)
    >>> assert set(d) == {x}
    >>> d[x]
    (oo > x, x > Max(z, y - 3*z - 4, y - 2*z + 2, 3*y/2 - z/2 - 1/2))
```

TiloRC commented:
> I'm having doubts that this even belongs in SymPy

The benefit to having a linear programing function within SymPy would be to get exact solutions to LP problems. If this sort of function doesn't belong in SymPy then where would it belong instead?

I noticed there's currently a [discussion](https://mail.google.com/mail/u/1/?ogbl#inbox/FMfcgzGtwMdQgSLhDlPFTBlghnXtzxnK) in the mailing list about being able to call NymPy functions using SymPy numbers. I wonder if there might be a way to call SciPy functions using SymPy floats. Then people would be able to use SciPy's linear programing function on SymPy rationals. 

> The simplex method added here is strictly for concrete matrices with Float/Rational args.

In [this comment](https://github.com/sympy/sympy/pull/25277#issuecomment-1630607985) Oscar mentioned that eventually you could implement this using DomainMatrix and handle irrational numbers sometimes. 

> Solving systems using the Fourier-Motzkin approach that was part of the original https://github.com/sympy/sympy/pull/22389 seems more SymPy apropos.

Could you use this to solve a LR problem? If so, would there be a big difference in speed?

smichr committed:
add lpmin, lpmax

smichr committed:
docstrings; unix EOL

smichr committed:
edits and helpers

smichr commented:
There was a concern raise [here](https://github.com/sympy/sympy/pull/25277#pullrequestreview-1551653796) about not having the function require that symbols of interest be passed. I am not convinced that this is necessary and find it annoying to have to provide information that the symbolically capable SymPy can determine directly. Here, as opposed to scipy, symbolic input can be provided. When it is, and since we are only dealing with linear systems *in symbols* (not arbitrary functions which may not be linear), the symbols are easily determined by `free_symbol` introspection of the objective function. This is similar to the way that `dsolve` and `solve` can figure out what to do given only an expression as in `dsolve(f(x).diff(x)-f(x))` or `solve(x - 2)`.

If someone wants to solve a linear system expressed in terms of `f(x), g(x), etc...` they will have to replace those functions with symbols, e.g.

![image](https://github.com/sympy/sympy/assets/90703/82868854-af9e-484e-b5a0-fc726903f2d5)

or work with the matrix representation of the system (which all functions added here will accept).

smichr commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,818 @@
+from sympy.assumptions.relation.binrel import AppliedBinaryRelation
+from sympy.functions.elementary.complexes import sign
+from sympy.matrices.dense import Matrix
+from sympy.matrices.immutable import ImmutableDenseMatrix
```
```suggestion
from sympy.matrices.matrices import MatrixBase
```

smichr commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,818 @@
+    if not args or len(args) > 4:
+        raise ValueError('expecting 1 - 4 args')
+    LP = [sympify(i) for i in args]
+    if len(LP) > 2 and all(isinstance(i, (list, ImmutableDenseMatrix)) for i in LP):
```
```suggestion
    if len(LP) > 2 and all(isinstance(i, (list, MatrixBase)) for i in LP):
```

smichr committed:
test for more general MatrixBase

sympification of Matrix gives ImmutableMatrix; input might also be SpareMatrix. In either case these are instances of MatrixBase.

smichr committed:
watch out for True or False conditions in primal_dual

smichr commented:
@TiloRC , could you look this over, please?

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+    replacements of symbols as given in r, {xi: expression with Xj},
+    so all variables in x+X will take on nonnegative values.
+
+    If symbols not in objective will be set to 0.
```
What if:

Minimize x subject to y = 2,  x=y

If we get rid of y, then the constraint x=y becomes meaningless.

smichr replied:
That would require resolving equalities, favoring those that don't appear in the objective. I preferred that, anyway, and can look at making that change.

TiloRC replied:
The example I gave involved equalities, but constraints involving variables not in the objective function can also be meaningful in situations without equalities: 

Minimize x subject to y >= 10, x >= y

Here the problem becomes unbounded when you remove y even though originally this wasn't the case.

smichr replied:
This still doesn't work: `lpmin(x,[y >= 10, x >= y])` -> Infeasible

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+
+    how = str(min_max).lower()
+    if 'max' in how:
+        # _simplex minimizes for Ax <= B so we
+        # have to change the sign of the function
+        # and negate the optimal value returned
+        _o, p, d = _simplex(A, B, -C, -D)
+        o = -_o
```
Are you sure this works?

I would think that flipping the sign of the objective function and then flipping the sign of the optimum value wouldn't work because of the constraints.

Wouldn't something along these lines happen?

maximize x + y subject to 
x <= 2
y <= 3

becomes

minimize x + y
subject to 
x <= 2
y <= 3

which is unbounded even though the original problem was not.

TiloRC replied:
Oh I see _lp_args handles this I think.

TiloRC replied:
I see that it's necessary to keep min_max as an input to this function so that the sign of the final output can be flipped. Couldn't the flipping of the sign of f take place in _lp_args?

TiloRC replied:
I think it might be better to handle all of the logic for converting the maximization problem to a minimization problem and back within `lpmax`. 

The reduction from maximization problem to minimization problem should be very clear.

TiloRC replied:
To make the reduction clear I think it should take place together, not split up over multiple functions.

smichr replied:
> Couldn't the flipping of the sign of f take place in _lp_args?

`_lp_args` returns f and const in standard from from the matrices. It is only because `_simplex` doesn't solve the standard form that the handling specifically for `_simplex` is handled at the point of call. That way any other simplex solver could be implemented and the special handling would happen only at the point of call.

smichr replied:
> To make the reduction clear I think it should take place together, not split up over multiple functions.

This is the flow:
```
lpmin(*args)
    args -> interpreted into standard f, constr in _lp_args
        f, constr -> matrix with aux variables in _lp_matrices
            A,B,C,D -> solution' (perhaps in terms of aux) in _lp
                solution' -> final solution in lpmin
```

smichr replied:
Perhaps we make lpmin only accept f, constr. If the user wants to use matrices then the call would be 
* `lpmin(*primal_dual(M)[0])`  if M is for f and `Ax<=b` or 
* `lpmin(*primal_dual(M.T)[1])` if M is for f and `Ax>=b` or 
* `lpmin(*primal_dual(_m(a,b,c,d))[0])` if a,b,c,d are the parts from `f=cx-d` and `ax<=b`. 
* etc...

In other words, give them the tools for changing from one form to the other and let them use them or create lambda's as they desire.

TiloRC replied:
> If the user wants to use matrices

One of the reasons to use matrices is efficiency, so converting matrices to equations would defeat the purpose.

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+    return how, f, ineq
+
+
+def lpmin(*args):
+    """return minimization of linear equation ``f = c*x - d``
+    under constraints ``A*x >= b``.
+
+    Input can be the ``Matrix([[A, b], [c, d]])``, the individual
+    matrices ``A, b, c, d`` or the function ``f`` and a list of
+    constraints expressed using Ge, Le or Eq.
```
I'm not sure if I like the idea that this can take multiple types of input. I think this might make it a little confusing to use. Are there other functions in SymPy that can do something like this?

smichr replied:
python `range` does this. SymPy often works with `*args` and `args` and handles conversion. @oscarbenjamin point out that "Mathematica's LinearOptimisation function has the signature LinearOptimisation(objective, constraints, vars) but can alternatively accept matrices like LinearOptimisation(f_vec, (c_mat, b_vec))." When the logic is straightforward for dealing with the different inputs, I prefer the single interface. But if sentiment is strongly against it, we could do `lpmin_from_matrix`, `lpmin_from_matrices` in addition to `lpmin` But this will not make the viewed code any more transparent to someone that knows the different outputs: it will be `lpmin(M)` (one arg), `lpmin(f, constr)` (two args) or `lpmin(a, b, c)` or `lpmin(a,p,c,d)`.

TiloRC replied:
I think it's reasonable to it this way then. Having multiple functions for the same thing does seem annoying. 

I would maybe mention that the input can be a list constraints expressed using Ge, Le or Eq first, before the matrix stuff. I think most people would prefer not to use the matrix. 

What's the reasoning for accepting either M or A, b, c, d? In what situation would someone prefer to use one or the other. It seems simpler to me to either accept just M or A, b, c, d.

smichr replied:
If someone is making A,b,c,d from linear_eq_to_matrix then it's not too hard to wrap it up before sending it `lpmin(Matrix([[A,b],[c,d]]))` but I got that wrong enough times that I just didn't want to think about it -- I know I have A,b,c,d right and just want the function to put them in M correctly.

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+
+def lpmin(*args):
+    """return minimization of linear equation ``f = c*x - d``
+    under constraints ``A*x >= b``.
```
probably good idea to explicitly mention the constraints must be linear just to be super clear

under constraints ``A*x >= b``

--->

under linear constraints ``A*x >= b``

smichr replied:
I can add "for the vector of symbols, ``x``." Then, by definition, f and constraints are linear.

TiloRC replied:
I don't think this is very important either way. Do whatever you think is best.

TiloRC commented:
I'm not sure if I like the `primal_dual` function being public. The particular way that LP are represented as matrices in our code is not universal and I don't think it's the most common way to do it either. What's the reasoning for including it?

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+    return how, f, ineq
+
+
+def lpmin(*args):
```
Instead of *args, the parameters could be 

objective, constraints 

where 

objective : either an expression or Matrix([c, d])

constraints : either a list of  constraints expressed using Ge, Le or Eq or Matrix([[A, b]])

smichr replied:
See below: there can be 1 (matrix), 2 (f, constr), 3 or 4 args (individual matrix components).

TiloRC commented:
Perhaps `lpmin` and `_lp` could be combined into one function. And then `lpmax` could its problem into minimization form, pass that along to `lpmin` and then modify the output before returning it.

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+    3 args: 3 lists/matrices: A, B, C, D=0
+    4 args: 4 lists/matrices: A, B, C, D
+
+    return: ('max' or 'min', objective, constraints)
```
So the returned objective function is an expression and constraints contain a list of Ge, Le, Eq?

It was my understanding that the purpose of being able to handle matrices was for efficiency reasons.  Wouldn't converting large matrices into SymPy expressions and then back into matrices be slow?

Why not have _lp_args convert everything to a standard matrix form?

smichr replied:
I'll look at refactoring. It might take some time given the current schedule.

smichr replied:
Now that I am not solving equalities, I could stay in "matrix mode" in `linprog`. But that might take another week to get time enough to finish it.

TiloRC commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,821 @@
+    unbound = set()  # symbols designated as unbound
+
+    for i in constr:
+        if isinstance(i, AppliedBinaryRelation):
```
I don't think it's necessary to handle predicate from the new assumptions anymore. We can get rid of this.

smichr commented:
> The particular way that LP are represented as matrices in our code is not universal

I thought Ferguson's use of "standard minimization/maximization" made the matrix unambiguous. I like it as a way to see the symbolics of the matrix that is of interest. I do not yet "think in matrices" when it comes to visualization. I know what `y <= x + 1` looks like but `[-1, 1, 1]` is less transparent to me. (One could also plot expressions, thus generated, in low dimensions.)

smichr commented:
> Perhaps `lpmin` and `_lp` could be combined into one function. And then `lpmax` could its problem into minimization form, pass that along to `lpmin` and then modify the output before returning it.

I am thinking of

* lp solves the standard minimization in matrix form with scipy signature
* lpmin/lpmax and helpers converts input to matrix form and pass to lp

If someone wants to work in matrix form they have to do the sign wrangling on their own and pass the matrices to lp.

oscarbenjamin commented:
> > Perhaps `lpmin` and `_lp` could be combined into one function. And then `lpmax` could its problem into minimization form, pass that along to `lpmin` and then modify the output before returning it.
> 
> I am thinking of
> 
> * lp solves the standard minimization in matrix form with scipy signature
> * lpmin/lpmax and helpers converts input to matrix form and pass to lp

I think that makes sense. I would call it `linprog` though which is the name used by both SciPy and Matlab.

The lpmin/lpmax functions are analogous to Mathematica's `LinearOptimization`. I prefer a name that clearly has min or max though. It could be `linprog_max` or `linear_maximize`, `maximize_linear`. There is already a function `maximum` that t would need to be distinct from. Also there is `linsolve` and so you could imagine `linmax` by analogy or maybe that's just confusing.

smichr committed:
use linsolve

smichr committed:
add linprog

smichr committed:
solve equalities

smichr committed:
remove AppliedBinarRelation handling

smichr committed:
problems with Eq handling

smichr commented:
I added a `linprog` routine. For lack of time I am falling back to constructing the equations to derive the matrices. I probably only need to do so if there are equalities and non-default ranges, but for now it all gets the same treatment.

I do not know what is wrong with the Eq handling, however. There are some XFAIL tests and an annotated doctest that show the bad results.

TiloRC commented:
> The particular way that LP are represented as matrices in our code is not universal

What I meant by this is that usually A is an m by n matrix where m is the number of slack variables (and there's one slack variable for each constraint) while n is the number of slack plus non-slack variables. For example:

```
x + y <= 2
2x + 3y <= 3
--->
x + y + s1 = 2
2x + 3y + s2 = 3
```
Normally, you would get something like this:
```
>>> A
Matrix([[1, 1, 1, 0],
              [2, 3, 0, 1]]) 
```
However, in our code this would be represented like:
```
>>> A
Matrix([[1, 1],
              [2, 3]) 
```
 
As a result of this difference, the pivot operation works a bit differently. Normally, pivoting performs a series of row operations. This means that the row reduced echelon form of the result of pivoting is the same as the row reduced echelon form of the original matrix. However, Ferguson's pivot operation doesn't have this property. Perhaps Ferguson's method is a bit faster/more memory efficient because the matrices don't have to be as big. However, it does seem nonstandard.

smichr committed:
add 2 equations for equalities

TiloRC commented:
>  Maybe I should just add the L - R <= 0 and R - L <= 0 conditions for Eq(L, R).

I think this should be fine. I'm not sure avoiding this will even be faster in the first place.

smichr committed:
make primal_dual private

smichr commented:
`primal_dual` is now private and the issues with Eq are resolved by adding the two equations.

smichr committed:
add tests for linprog

smichr committed:
add linear clarification to docstring

smichr committed:
load all symbols into matrix

smichr commented:
I believe everything is working now.

smichr commented:
@TiloRC, @oscarbenjamin a question I have is how to interpret the matrix entries. `_simplex` automatically assumes that all variables will be non-negative so 
```python
>>> _simplex([1],[3],[1],[0])  # optimize.linprog([[1]],[[1]],[3])
(0, [0], [0])
```
But the way I have been interpreting a condition like `x <= 3` is that this represents an semi-unbounded constraint. Since sympy works with the symbolic inequalities I think this is a plausible interpretation (in absence of a `x >=0` constraint, too). But when given a matrix to interpret I suspect that this could cause confusion since a row `[1, 0, ..., 0, 3]` for `x <= 3` when passed to the `_simplex` method *automatically* treats `x` as nonnegative. 

So to avoid confusion for anyone familiar with scipy, I think it should be made clear that you cannot interchange between matrices and equation (e.g. matrix -> inequalities -> matrix*). The code needs to be corrected for any inconsistencies in that, too.

TiloRC commented:
I'm not sure I totally understand what you're saying. Are you concerned that people might get confused when they use a matrix form because they don't expect the variables to have a nonnegative constraint by default?

SciPy by default also automatically assumes each variable is nonnegative—which is very confusing because the documentation doesn't make this very clear. However, it also offers the option to set whatever boundaries you like on all the variables or only some of them.

smichr committed:
lpmin/max for symbolics; quick exit for linprog

smichr commented:
> Are you concerned that people might get confused when they use a matrix form because they don't expect the variables to have a nonnegative constraint by default?

Yes. I changed the docstrings to try make this clear.

smichr committed:
reorganize simplex.py

smichr commented:
I think this is ready for a last view. The `linprog` now only reconstructs symbolic systems to deal with bounds and equalities because I haven't taken the time to figure out how to modify the matrices directly. Here is a comparison of calling it vs lpmin:
```python
>>> lpmin(x + y, [x -2*y<=-6, x<=3,x>=1])
(9/2, {x: 1, y: 7/2})
>>> a,b,c,d,j,j,j=_lp_matrices(x + y, [x -2*y<=-6, x<=3,x>=1]) # extra variable introduced to handle x bounds
>>> linprog((c,d),a,b)
(9/2, {x1: 0, x2: 7/2, x3: 0})

Handle bounds without relationals:

>>> a,b,c,d,j,j,j=_lp_matrices(x + y, [x -2*y<=-6])
>>> linprog((c,d),a,b,bounds=[(1,3),(0,None)])
(9/2, {x1: 1, x2: 7/2})
```

smichr committed:
handle Eq in linprog

smichr committed:
linprog does not use intermediate expr

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+    (0, 4), (0, 0) and (2, 0). The minimum of ``f`` occurs at (2, 0).
+    Define A, B, C, D for the standard minimization:
+
+    >>> A, B, C, D = [Matrix(i) for i in [[[2, 1]], [4], [[1, 1]], [-1]]]
```
I think it would be clearer to state these explicitly like `A = Matrix([[2, 1]])`, `B = Matrix(...)` rather than using a list comprehension.

oscarbenjamin replied:
Maybe this could even show the inequalities to clarify the relationship between the matrices and the problem:
```python
In [5]: M = Matrix([[1, 2], [3, 4]])

In [6]: x = Matrix(symbols('x:2'))

In [7]: b = symbols('y:2')

In [8]: [a >= b for a, b in zip(M*x, b)] # No way to avoid a list comprehension here...
Out[8]: [x₀ + 2⋅x₁ ≥ y₀, 3⋅x₀ + 4⋅x₁ ≥ y₁]
```

smichr replied:
done

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+    if A and B:
+        M = Matrix([[A, B], [C, D]])
+    else:
+        assert not A and not B  # no constraints
```
I think this assert is guaranteed by the immediately preceding if (no big deal)

oscarbenjamin replied:
Oh, no it isn't. How do we know this is satisfied?

smichr replied:
It now raises a ValueError

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+        piv_cols = [_ for _ in range(n) if C[_] < 0]
+        if not piv_cols:
+            break
+        c = sorted(piv_cols, key=lambda _: X[_])[0] # Bland's rule
```
I guess this could be `min` or `max` rather than `sorted`/

smichr replied:
done

smichr replied:
Note: I actually struggled with how to say this. I know about the decorated sort but the "decorated max" was a little more ellusive. I got it back to a one-liner, however.

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+
+    >>> from sympy import Matrix
+    >>> from sympy.solvers.simplex import _abcd
+    >>> m = Matrix(3, 3, range(9))
+    >>> L = _abcd(m, list=True); L
+    ([[0, 1], [3, 4]], [2, 5], [[6, 7]], [8])
+    >>> _abcd(m)
+    (Matrix([
+    [0, 1],
+    [3, 4]]), Matrix([
+    [2],
+    [5]]), Matrix([[6, 7]]), Matrix([[8]]))
+    >>> assert tuple(Matrix(i) for i in L) == _
```
I think many people would struggle to understand this when reading it. How about:
```
>>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
>>> M
Matrix([
[1, 2, 3],
[4, 5, 6],
[7, 8, 9]])
>>> A = M[:2,:2]
>>> A
Matrix([
[1, 2],
[4, 5]])
>>> B = M[2,:2]
>>> B
Matrix([[7, 8]])
```
etc. To me seeing each result and comparing it with the original is much clearer than interpreting the code.

smichr replied:
modified

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+        if i == False:
+            return  # no solution
+        i = i.canonical  # +/-oo, if present will be on the rhs
+        if isinstance(i, (Le, Ge, Lt, Gt)) and i.rhs.is_infinite:
+            if isinstance(i, (Lt, Le)) and i.rhs is S.NegativeInfinity:
+                return False
+            if isinstance(i, (Gt, Ge)) and i.rhs is S.Infinity:
+                return False
```
Maybe it better to just raise an error if any side is infinite...

smichr replied:
There is no other way to pass an inequality that allows for any real value. `linprog` uses tuples for limits with None acting as infinity.

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+    return o, p
+
+
+def lpmin(f, constr):
+    """return minimum of linear equation ``f`` under
+    linear constraints expressed using Ge, Le or Eq.
+
+    All variables will take on nonnegative values unless
+    a univariate constraint indicates other ranges for
+    a symbol, e.g, ``x <= 3`` will allow for negative
```
Is it simple to avoid assuming nonnegativity by default?

smichr replied:
It is not hard to do so. It is harder to think what will make the most sense to people using the function. The inequality `x <= 3` clearly permits negative values. So I would rather that the user also indicate x >= 0 if they also want that condition. If I automatically add `x >= 0` then how will the user indicate a range like -1 to 1?

oscarbenjamin replied:
> It is harder to think what will make the most sense to people using the function

I think most users would assume that since they are providing assumptions the function will use the provided assumptions rather than imposing any assumptions a priori.

smichr replied:
Re-reading this, I think I miscued on your intent. I think you are saying that the variables should be unbounded unless constrained otherwise. This is as simple as adding `xi <= oo` for all `xi`. If they want nonnegative values then they should explicitly say `xi >= 0`.

oscarbenjamin replied:
I would expect the method not to assume anything other than that symbols are real by default. It shouldn't be necessary to pass `xi <= oo` in order to specify that `xi` might be negative.

smichr replied:
This, now, is how the routine behaves as mentioned in my comment (way below). I totally agree.

oscarbenjamin commented on sympy/solvers/simplex.py:
```diff
@@ -0,0 +1,966 @@
+        b = b.col_join(b_eq)
+        b = b.col_join(-b_eq)
+
+    ## the objective
+    if isinstance(cd, MatrixBase):
+        C, D = cd, Matrix([0])
+    elif len(cd) == 2 and all(isinstance(i, (MatrixBase, list)
+            ) for i in cd):
+        C, D = [Matrix(i) for i in cd]
+    elif isinstance(cd, list) and isinstance(cd[0], list) and len(
+            cd[0]) == A.cols:
+        C, D = Matrix(cd), Matrix([0])
+    else:
+        raise ValueError('expecting one or two matrices/lists for cd')
```
Is it actually useful to allow these different possibilities? I find it hard to follow what this function does because of the ambiguity about its arguments.

oscarbenjamin replied:
Especially for a public function we need to be careful about signature. I don't see a good reason to stray from what SciPy and Matlab do in this function signature.

oscarbenjamin replied:
This is my only real concern for this PR.

smichr replied:
Anyone using scipy can (up to the bounds) pass the same variables there as here. The only difference is that this `linprog` allows a `D` -- but since that doesn't factor into the solution process we could leave it off and let the user ponder why it is not needed. For simplicity I am not opposed to omitting `D`.

oscarbenjamin replied:
I should take a closer look to understand how these APIs work.

No objection from my side but I thought I understood this and probably I didn't...

smichr replied:
It has now been omitted.

smichr replied:
> No objection from my side but I thought I understood this and probably I didn't..

I am very appreciative of the time given for review. I agree that when working with symbolic constraints that we should let the user make the assumptions rather than imposing an assumption made by a support routine. So now `lpmin/max` are WYSIWYG in terms of constraints and infinities are disallowed (because those are part of the default assumption for variables). See the simple 1-D example in the OP.

The basic idea is that a linear function can be minimized on a simplex (n-dim linear region). `_simplex` does the task of minimizing `C*x - D` with `A*x <= B` for nonnegative variables, `x`. Everything else is scaffolding to get a system expressed with Le, Ge or Eq and desired bounds expressed in terms of A, B, C, D and to handle finding the max or the min.

smichr committed:
increase coverage

smichr committed:
modifications post review

smichr commented:
@TiloRC - I addressed all @oscarbenjamin concerns, do you want to give it a final once over?

smichr committed:
check for oscillation

smichr committed:
lpmin,lpmax default to unbounded

smichr committed:
black edits

smichr commented:
[ ] fix handling of bounds in linprog
```python
>>> x = x1, x2, x3, x4 = symbols('x1:5')
>>> X = Matrix(x)
>>> M=linear_eq_to_matrix
>>> c, d = M(5*x2 + x3 + 4*x4 - x1, x)
>>> a, b = M([5*x2 + 2*x3 + 5*x4 - (x1 + 5)], x)
>>> aeq, beq = M([Eq(3*x2 + x4, 2), Eq(-x1 + x3 + 2*x4, 1)], x)
>>> linprog(c, a, b, aeq, beq)
(9/2, [0, 1/2, 0, 1/2])
>>> from sympy.solvers.simplex import *
>>> show_linprog(c, a, b, aeq, beq, bounds=(0,1))  # constraints shouldn't break the solution
(-x1 + 5*x2 + x3 + 4*x4, [-x1 + 5*x2 + 2*x3 + 5*x4 <= 5, Eq(3*x2 + x4, 2), Eq(-x
1 + x3 + 2*x4, 1), x1 >= 0, x1 <= 1, x2 >= 0, x2 <= 1, x3 >= 0, x3 <= 1, x4 >= 0, x4 <= 1])
>>>
>>> lpmin(*_)
(9/2, {x1: 0, x2: 1/2, x3: 0, x4: 1/2})
>>> linprog(c, a, b, aeq, beq, bounds=(0,1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\leslie\sympy\sympy\solvers\simplex.py", line 1071, in linprog
    o, p, d = _simplex(A, b, C)
  File "C:\Users\leslie\sympy\sympy\solvers\simplex.py", line 345, in _simplex
    raise InfeasibleLPError(filldedent("""
sympy.solvers.simplex.InfeasibleLPError:
Oscillation: this can happen when there are variables in the
constraints (A) that are not in the objective (C). If you think there
is a solution for this case please report this as a bug.
```

oscarbenjamin commented:
I think that in the end the functions here should be exposed at top-level like
```
from sympy import linprog
```

oscarbenjamin commented:
This is nice:
```python
In [8]: from sympy.solvers.simplex import lpmin

In [9]: lpmin(x+y, [x>=2, y>=2])
Out[9]: (4, {x: 2, y: 2})
```

oscarbenjamin commented:
Playing around with this I get the error message
```
TypeError: only Ge or Le is allowed, not Ne(y, 2)
```
but actually Eq is allowed:
```
In [32]: lpmin(x+y, [x>=-10, Eq(y, 2)])
Out[32]: (-8, {x: -10, y: 2})
```
Also perhaps Ge and Le might not be intuitive to users so something like "only equalities and non-strict inequalities are allowed".

oscarbenjamin commented:
> ```
> 
> ```
> 
> 
>     
>   
> 
> Also perhaps Ge and Le might not be intuitive to users so something like "only equalities and non-strict inequalities are allowed".

Or even more explicit "only equalities like Eq(x, y) or non-strict inequalities like x >= y are allowed in lp",

oscarbenjamin commented:
@smichr this is still marked as WIP. Do you intend for this to be in ready to merge or is there more work do do?

smichr commented:
> Do you intend for this to be in ready to merge

I identified an issuer [here](https://github.com/sympy/sympy/pull/25440#issuecomment-1676049751). I can, apparently, get the right form when dealing with expressions but when trying to recreate the matrices directly I use a slightly different approach and I am not sure why it fails. The system is right but how it is presented to the solver leads to oscillations while checking feasibility. At this point I am not sure how to do directly what I do using a replacement dictionary. In one case I let an unbound variable x be represented as x - u (re-using the existing variable and identifying the remaping x-> x  - u and update the objective and other constraints accordingly. In the failing case I simply add a new equation x = v - u and two new variables and add the two lines to the matrix for that equalit, corresponding to  `x -  v + u <= 0` and `-(x - v + u) <= 0` and don't make any replacements.

smichr committed:
show_linprog

smichr committed:
correct padding and detect oscillations

smichr committed:
correct reference

smichr committed:
add ref for cycling

smichr committed:
expose to * import

smichr committed:
check both solutions just to be safe

smichr commented:
I stop when an ABA cycle is detected and check the answer at the end: if the solution is not valid, an error raises. Ferguson's paper suggests that cycling should be avoided with Bland's rule, but the `# o2` example I give oscillates yet will still reach the solution if phase two is allowed to continue.

smichr commented:
Could someone who uses documentation check to see that my additions look right, please?

smichr commented:
OK, I will commit this as is and see if I can work out something better for the oscillation. Perhaps I can get some feedback from any users in the meantime.

TiloRC commented:
Are the random problems that involve scipy getting run by the optional dependencies?

If I recall correctly, I think inequalities should be added to this list https://github.com/sympy/sympy/blob/master/bin/test_optional_dependencies.py#L20-L78 so that the optional dependencies can run.
